lib/models/addon.js

```diff
@@ -29,13 +31,59 @@ var ensurePosixPath = require('ensure-posix-path');
 var defaultsDeep = require('ember-cli-lodash-subset').defaultsDeep;
 var Babel = require('broccoli-babel-transpiler');
 var findAddonByName = require('../utilities/find-addon-by-name');
+var experiments = require('../experiments');
+var heimdall = require('heimdalljs');
+var calculateCacheKeyForTree = require('../utilities/addon-calculate-cache-key-for-tree');
+
+heimdall.registerMonitor('addon-tree-cache', function AddonTreeCacheSchema() {
+  this.hits = 0;
+  this.misses = 0;
+  this.adds = 0;
+});
 
 var DEFAULT_BABEL_CONFIG = {
   modules: 'amdStrict',
   moduleIds: true,
   resolveModuleSource: require('amd-name-resolver').moduleResolve
 };
 
+var DEFAULT_TREE_FOR_METHODS = {
```

Comment:
Note that it also invalidates the caching strategy not for itself, but for `eachAddonInvoke`d. Whatever gets invoked could have side effects that change how something gets built. I'd like for us to try and find a way to warn when we catch this occurring.